src/lib.rs

```diff
@@ -126,15 +126,49 @@ pub trait Renderable: Send + Sync {
     fn render(&self, context: &mut Context) -> Result<Option<String>>;
 }
 
+pub type TemplateLocation = Path;
+pub trait FileRepository {
+    fn read_template_file(&self, path: &TemplateLocation) -> Result<String>;
```

Comment:
Sure! I think there's one missing concept right now in liquid-rust. Let's ignore that for now and see how `TemplateFactory` and `TemplateRepository` are used together. 

```rust
// This may not be idiomatic Rust, sorry for that!

// The TemplateRepository knows how to find the template content 
// either on the disk, in the database, on S3, whatever.
trait TemplateRepository {
  fn fetch(template_name: TemplateName) -> Option<Result<String>>;
}

// This is the current implementation in liquid-rust
// It uses a basic default path, and naively appends
// the template name.
struct LocalTemplateRepository {
  root: PathBuf,
}

impl TemplateRepository for LocalTemplateRepository {
  fn fetch(template_name: TemplateName) -> Option<Result<String>> {
    // read from file system
  }
}

// We could also provide a default caching repository.
// If it's cache doesn't contain an entry for the template,
// it would use it's backend to fetch it, then grab ownership
// of the template, and lend it when a template needs to be 
// rendered.
// I'm not very versed in lifetimes (haven't had to deal with them
// yet, sorry again :)
struct<'a> TemplateRepositoryCache<'a> {
  cache: MapOfSomeSort<TemplateName, &'a Template>,
  backend: Box<TemplateRepository>,
}

impl TemplateRepository for TemplateRepositoryCache {
  cache.entry(template).or_insert_with(|| self.backend.generate(...))
}

// You could also have a TemplateRepository that uses multiple
// other repositories, falling back to the others if the template
// is not found.
struct CascadingTemplateRepository {
  backends: Vec<Box<TemplateRepository>>,
}

// The TemplateFactory is the equivalent of today's `parse` function. 
// It takes a repository to find the template contents, and returns the 
// templates themselves. 
trait TemplateFactory {
  pub fn generate(&self, template_name: TemplateName, options: LiquidOptions) -> Option<Result<&Template>>
}

// The only difference with the current way is that it uses a repository
// to locate the template contents
struct DefaultTemplateFactory {
  template_repository: Box<TemplateRepository>,
};

impl TemplateFactory for DefaultTemplateFactory {
  pub fn generate(...) -> Options<Result<&Template>> {
    let template_content = self.template_repository.fetch(template_name);
    parse(template_content, options)
  } 
}

// Same as for the repository, the factory can have a naive cache.
struct TemplateFactoryCache {
  cache: MapOfSomeSort,
  backend: Box<TemplateFactory>,
}

impl TemplateFactory for TemplateFactoryCache {
  pub fn generate(...) -> Option<Result<&Template>> {
    cache.entry(template).or_insert_with(|| self.backend.generate(...))
  }
}

let repository = S3TemplateRepository::new(my_s3_bucket);
let factory = DefaultTemplateFactory::new(repository);
let options = LiquidOptions::new(factory, ...);
let template = factory.generate(template_name, options);
let output = template.render(&mut context);
```

The problem I see with this isn't a dealbreaker, but it bugs me that the factory needs to know about the options, and the options about the factory. That's where the `Renderer` comes in.

```rust
let liquid_renderer = LiquidRenderer::new(factory, options);
let output = liquid_renderer.render(template_name, &mut context);
```